src/sched/lthread.c

```diff
@@ -233,13 +231,31 @@ void lthread_run(void)
             dequeued = 0;
             if (mpmc_dequeue(&__scheduler_queue, (void**)&lt))
             {
+                SGXLKL_ASSERT(!(lt->attr.state & BIT(LT_ST_EXITED)));
+
                 dequeued++;
                 pauses = sleepspins;
-                a_dec(&schedqueuelen);
                 SGXLKL_TRACE_THREAD(
                     "[%4d] lthread_run(): lthread_resume (dequeue)\n",
                     lt ? lt->tid : -1);
                 _lthread_resume(lt);
```

Comment:
Maybe skip non-kernel threads here if `_lthread_terminating_scheduler == sched`?

Reply:
Hmm, couldn't this lead to deadlock in the LKL scheduler if it wants to schedule them?

Reply:
Yes, possibly. Does LKL allow syscalls to enter the kernel after shutdown has started? It probably shouldn’t.

Reply:
Currently we don't: https://github.com/lsds/lkl/compare/385f721b339fe48b188b4924c2663e1ea2cdeb13...8552da7975f1db1e178b802a28409ea2c6d12aab#diff-040a3c76c7eac15c7395c8f8a408af4e

What error should we return to the userspace thread when it's issuing a syscall? I'm worried that this would result in weird errors in the application code. It would be better to also deallocate the userspace thread. I still think that the kernel should be killing userspace threads during shutdown. I don't understand why not all threads get killed (which also leads to the problem in `lkl_sys_halt()` -- see the comment there.)

Perhaps we can try this a part of another PR, as this one is already getting too big?

Reply:
This may work, but let's try this in another PR.

Reply:
I've now added a guard that ensures that syscalls during shutdown can only come from the termination thread.